Sprint 5

O-o-l-e-g opened the pull request.

KonstantinChikishevWG committed:
- Fixing lenta controller

grigorii-sum committed:
Добавлен endpoint для отправки сообщений в Kafka
Изменена структура пакета "parser" (добавлены интерфейсы и в них вынесено описание классов и методов)
Добавлены логи

grigorii-sum committed:
Merge branch 'sprint-5' of https://github.com/microservices-course-itmo/lenta-parser-service into sprint-5

# Conflicts:
#	src/main/java/com/wine/to/up/lenta/service/controller/LentaStoreController.java
#	src/main/java/com/wine/to/up/lenta/service/parser/ParserRsp.java

grigorii-sum committed:
refactoring

butenko-or committed:
Change for checking deploy

KonstantinChikishevWG committed:
- Replacing UpdatingProducts to ParserApi

viprikz committed:
- Adding Tests

kosch71 committed:
- Fixing job
- Fixing kafka controller
- Refactoring ProductDTO
- Tiny fix in EventSerializer
- Wait while Eugeniy complete parser, then i will LentaWineParserServiceImpl

kosch71 committed:
- Adding checking in ExportProductDtoList and SendKafkaController while we wait completed parser
- Deleting some build properties in LentaWineParserServiceImpl
- Changing LentaStoreController Response

kosch71 committed:
- Move Tests

kosch71 committed:
- Adding Selenium in pom.xml
- Refactoring LentaWineParserServiceImpl(Adding new properties)
- Adding new property in lenta-site.properties

eomelyanovich committed:
Изменение парсера: парсинг новых пропертей и драйвер для селениума

grigorii-sum committed:
Tests

grigorii-sum committed:
Merge remote-tracking branch 'origin/sprint-5' into sprint-5

eomelyanovich committed:
Refactoring parser: full parsing, changing parser config

kosch71 committed:
## Parser rsp integration
- Adding normal unpacking of json
- Adding DRY sugar and RED colour as default because unrecognized throw Ex.(common-parser-api)
- Refactoring properties in lenta-site.properties
- Uncommenting packing from our dto to api dto

viprikz committed:
Parser test

MaximKuznetsov17 commented on pom.xml:
```diff
@@ -68,16 +82,45 @@
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.12</version>
+        </dependency>
```
Зачем подключаете junit? Причём дважды
Эта библиотека есть в spring-boot-starter-test

MaximKuznetsov17 commented on src/main/java/com/wine/to/up/lenta/service/controller/LentaStoreController.java:
```diff
@@ -1,30 +1,27 @@
 
     @Autowired
-    private ParserReqService parserReqService;
+    private ParserReqServiceImpl parserReqServiceImpl;
```
Нужно избавляться от аннотаций Autowired. Рекомендую прописать конфигурацию ваших контроллеров в соответствующий пакет.

MaximKuznetsov17 commented on src/main/java/com/wine/to/up/lenta/service/controller/LentaStoreController.java:
```diff
@@ -1,30 +1,27 @@
-        return list;
+    @GetMapping(name = "/parser", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<Object> getParserResult() {
+        return new ResponseEntity<>(parserReqServiceImpl.getJsonList().getWineList().toList(), HttpStatus.OK);
```
Это в целом выглядит ужасно
А если условный `getJsonList()` вернёт null у вас всё упадёт

kosch71 replied:
Более подробнее можно? Вторую часть коммента понял, но первую не совсем.

MaximKuznetsov17 replied:
Такие записи в одну строку выглядят нечитаемо. Лучше то, что в качестве response передаёте вынести. Плюс лучше всё-таки парочку проверок сделать, что всё как надо работает

MaximKuznetsov17 commented on src/main/java/com/wine/to/up/lenta/service/controller/SendKafkaController.java:
```diff
@@ -0,0 +1,119 @@
+    private LentaWineParserServiceImpl parseService;
+
+    @Autowired
+    private KafkaMessageSender<ParserApi.WineParsedEvent> kafkaSendMessageService;
```
Повторюсь, нужна конфигурация контроллеров, убираем `@Autowired`

MaximKuznetsov17 commented on src/main/java/com/wine/to/up/lenta/service/cron/ExportProductDtoList.java:
```diff
@@ -24,12 +25,12 @@
     @Value("${string.for.cron}")
```
А вы точно знаете что такое cron? Я по названию этой проперти ничего не понял

MaximKuznetsov17 commented on src/main/java/com/wine/to/up/lenta/service/parser/impl/ParserReqServiceImpl.java:
```diff
@@ -0,0 +1,156 @@
+        this.baseUrl = "https://lenta.com";
+        return this;
+    }
+    @SneakyThrows
```
Эту аннотацию следует убрать. Нужно нормально обрабатывать ошибки и писать о них в лог. У вас же сейчас лог практически пустой. И это касается не только этого участка кода

MaximKuznetsov17 commented on src/main/java/com/wine/to/up/lenta/service/parser/impl/ParserReqServiceImpl.java:
```diff
@@ -0,0 +1,156 @@
+    @SneakyThrows
+    public ParserRspImpl getJsonList() {
+
+        System.setProperty("webdriver.chrome.driver", "/Lessons/Java/lenta-parser-service/src/main/resources/chromedriver.exe");
```
Зачем вам Selenium в целом? exe файлы на линуксовом сервере не выполнятся. Нужно указывать относительный путь к файлу, чтобы запустилось не только на вашей машине.

MaximKuznetsov17 commented on src/main/java/com/wine/to/up/lenta/service/parser/impl/ParserReqServiceImpl.java:
```diff
@@ -0,0 +1,156 @@
+                .uri(URI.create(apiUrl))
+                .header("Content-Type", "application/json")
+                .POST(HttpRequest.BodyPublishers.ofString(apiBody))
+                .build();
```
Судя по содержимому `apiBody` вы парсите только 20 вин.

MaximKuznetsov17 commented on src/main/resources/lenta-site.properties:
```diff
@@ -1,3 +1,4 @@
-string.for.cron=https://www.lenta.com
\ No newline at end of file
+site.base.url=https://lenta.com
+string.for.cron=https://www.lenta.com
```
Уже писал выше про неудачное название для проперти

MaximKuznetsov17 commented on src/test/java/AbstractTest.java:
```diff
@@ -0,0 +1,21 @@
+    protected void setUp() {
+        mvc = MockMvcBuilders.webAppContextSetup(webApplicationContext).build();
+    }
+}
```
Нужны отступы. Есть большие проблемы с кодстайлом в тестах.

MaximKuznetsov17 commented on src/test/java/com/wine/to/up/lenta/service/IntegrationTest.java:
```diff
@@ -0,0 +1,44 @@
+    private SendKafkaController sendKafkaController;
+
+    @Test
+    public void test1() throws Exception {
```
Название теста должно хотя бы отдалённо давать понять что проверяется в нём.

MaximKuznetsov17 commented on src/test/java/com/wine/to/up/lenta/service/IntegrationTest.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @Test
+    public void test2() throws Exception {
```
То же самое

MaximKuznetsov17 commented on src/main/java/com/wine/to/up/lenta/service/parser/impl/ParserReqServiceImpl.java:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    public JSONObject getProperties(WebDriver driver, JSONObject jsonString, JSONArray array, int i){
+        driver.get(baseUrl + array.getJSONObject(i).getString("skuUrl"));
```
То, что вы делаете с помощью Selenium можно сделать с помощью Jsoup. Если нужна помощь, напишите нам, сделаем всё возможное, чтобы помочь вам разобраться.

kosch71 replied:
Человек, который делал парсер сказал, что lazy load через selenium можно обойти. Пробовали также через jsoup, но безуспешно.

MaximKuznetsov17 replied:
Вам следовало просто посмотреть какие запросы и куда шлёт клиент. Вы же додумались слать пост запрос. Можно сказать, что это уже половина дела. 
Проанализируйте этот ответ. Там можно найти общее количество вин. Ну, и чисто исходя из логики наименования полей, которые вы шлёте в теле запроса можно догадаться куда подставить это общее число вин.

MaximKuznetsov17 replied:
И да, я может чего-то по коду не понял, но вы же даже с помощью selenium не обошли эту проблему. Она как была так и осталась

kosch71 replied:
Проблема решилась с селениумом. Но я сейчас попробовал с помощью jsoup сделать(по аналогии с вашим парсером) и у меня все заработало.

MaximKuznetsov17 replied:
Всё равно не понимаю как вы все вина парсите...

kosch71 replied:
Тут пока не все. Надо изменить лимит вин в поперти для запроса (убрать лимит в принципе) и все вина будут тогда парсятся. Кураторы об этом уведомлены.

MaximKuznetsov17 replied:
Хорошо)

kosch71 replied:
По словам Жени, человеку который делал парсер, все проперти были записаны в один элемент я являются по сути потомками одного элемента. Они могут быть расположены в рандомном порядке.

kosch71 replied:
Могу скинуть пруф видео, где парсятся все характеристики.

MaximKuznetsov17 replied:
> убрать лимит в принципе

Попробовал сейчас, только 24 вина вернулось.
Видимо всё-таки нужно нормальный лимит выставить
Или я что-то не так сделал?

MaximKuznetsov17 replied:
> Могу скинуть пруф видео, где парсятся все характеристики.

Так вопрос не по характеристикам, с характеристиками всё понятно. Они вроде как по коду нормально парсятся)

kosch71 replied:
> > убрать лимит в принципе
> 
> 
> 
> Попробовал сейчас, только 24 вина вернулось.
> 
> Видимо всё-таки нужно нормальный лимит выставить
> 
> Или я что-то не так сделал?

Не, все так

MaximKuznetsov17 commented on src/main/java/com/wine/to/up/lenta/service/controller/SendKafkaController.java:
```diff
@@ -0,0 +1,119 @@
+                .collect(Collectors.toList());
+
+        ParserApi.WineParsedEvent message = ParserApi.WineParsedEvent.newBuilder()
+                .setShopLink("https://www.lenta.com")
```
Это лучше бы в константу вынести, а в идеале вообще из пропертей брать.

MaximKuznetsov17 commented on src/main/java/com/wine/to/up/lenta/service/controller/SendKafkaController.java:
```diff
@@ -0,0 +1,119 @@
+    private KafkaMessageSender<ParserApi.WineParsedEvent> kafkaSendMessageService;
+
+    @GetMapping("/kafka")
+    public void sendKafkaMessage() {
```
У вас код этого метода аналогичен коду из job-ы, на сколько я мог понять. Я думаю, что лучше внутри этого метода вызвать job. Всё-таки принцип DRY не просто так придуман:)

kosch71 committed:
## Refactoring from Crossroad team
- Adding ControllerConfiguration
- Adding ExportProductDtoList in SendKafkaController
- Adding try/catch blocks in ParserReqServiceImpl
- Adding ColourTest, SugarTest
- Removing selenium support and adding jsoup parsing in ParserReqServiceImpl
- Adding SendKafkaControllerTest
- Rename tests in IntegrationTest
- Removing Currency
- Adding null checking for rating in ExportProductDtoList

viprikz committed:
Parser test

viprikz committed:
Merge remote-tracking branch 'origin/sprint-5' into sprint-5

kmosunoff commented on src/test/java/com/wine/to/up/lenta/service/controller/SendKafkaControllerTest.java:
```diff
@@ -0,0 +1,31 @@
+    //GET API test
+    @Test
+    public void sendKafkaMessage() {
+        String uri = "/lenta/kafka";
```
LentaStoreControllerTest и SendKafkaControllerTest отличаются по большому счёту только одной константой, возможно, лучше вынести общее в абстрактный класс. Обработка ошибок и логирование - туда же.

kmosunoff commented on src/test/java/com/wine/to/up/lenta/service/controller/LentaStoreControllerTest.java:
```diff
@@ -0,0 +1,26 @@
+    //GET API test
+    @Test
+    public void getProductsList() throws Exception {
+        String uri = "/lenta";
```
Запрос идёт на роут, который ничего не возвращает (/lenta вместо /lenta/parser)?
В тесте также стоит и распарсить небольшое количество вин, проверить не только статус ответа, но и успешность парсинга, например, посчитав различные характеристики (10 красных вин, 5 белых и т. п.)

kmosunoff commented on src/test/java/com/wine/to/up/lenta/service/db/constans/ColorTest.java:
```diff
@@ -0,0 +1,23 @@
+        Assert.assertEquals(ParserApi.Wine.Color.ROSE, Color.resolve("Розовый"));
+    }
+
+}
```
Полезно, но не забывайте о варианте, когда передаётся некорректная (неизвестная) строка. То же самое касается SugarTest

kmosunoff commented on src/test/java/com/wine/to/up/lenta/service/parser/impl/ParserReqServiceImlpTest.java:
```diff
@@ -0,0 +1,52 @@
+        assertEquals(verifiedJson.toString(), parserReqServiceImpl.changeToLocal().getProperties(driver, jsonAns, jsonArr,0).toString() );
+        //return new ResponseEntity<Object>(parserReqServiceImpl.changeToLocal().getProperties(driver, jsonAns, jsonArr,0).toString(), HttpStatus.OK);
+    }
+}
```
Выглядит, как локальный тест, но зачем-то упоминается ещё и селениум.

kmosunoff commented on src/main/java/com/wine/to/up/lenta/service/configuration/KafkaConfiguration.java:
```diff
@@ -1,11 +1,10 @@
 
-import com.wine.to.up.commonlib.messaging.KafkaMessageHandler;
 import com.wine.to.up.commonlib.messaging.KafkaMessageSender;
 import com.wine.to.up.lenta.service.components.LentaServiceMetricsCollector;
```
LentaServiceMetricsCollector, кстати, надо обновить, передавая имя сервиса в конструктор, как того требует родительский класс.

kosch71 committed:
## ParserReqServiceImpl
- Adding properties for socket
- Adding rest template

kosch71 committed:
## ParserReqServiceImpl
- Deleting test

kosch71 committed:
## ParserReqServiceImpl
- Adding parsing for another wines
- Adding sparkling in ProductDTO
- Adding new Color and Sugar elements
- Renaming some methods

kosch71 committed:
## ParserReqServiceImpl
- Minor changes

grigorii-sum committed:
Refactoring for sonar

viprikz committed:
ParserConfiguration fix

viprikz committed:
Parser test

kosch71 committed:
## Minor changes
- Adding maven dependency
- Rename test name

kosch71 committed:
## Minor changes
- Deleting old dependence

kosch71 committed:
## Minor changes
- Renaming test names
